src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -176,10 +183,37 @@ def prune_heads(self, heads):
         self.pruned_heads = self.pruned_heads.union(heads)
 
     def _attn(self, query, key, value, attention_mask=None, head_mask=None):
-        attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
-        if self.scale_attn_weights:
-            attn_weights = attn_weights / (float(value.size(-1)) ** 0.5)
+        if self.reorder_and_upcast_attn:
+            # Use `torch.baddbmm` (a bit more efficient w/ alpha param for scaling -- from Megatron-LM)
+            bsz, num_heads, seq_len, dk = query.size()
+
+            # Preallocate attn_weights for `baddbmm`
+            attn_weights = torch.empty(bsz * num_heads, seq_len, seq_len, dtype=torch.float32, device=query.device)
+
+            # Compute Scale Factor
+            scale_factor = 1.0
+            if self.scale_attn_weights:
+                scale_factor /= float(value.size(-1)) ** 0.5
+
+            if self.scale_attn_by_layer:
+                scale_factor /= float(self.layer_idx)
+
+            # Upcast (turn off autocast) and reorder (Scale K by 1 / root(dk))
+            with autocast(enabled=False):
+                q, k = query.reshape(-1, seq_len, dk), key.transpose(-1, -2).reshape(-1, dk, seq_len)
+                attn_weights = torch.baddbmm(attn_weights, q.float(), k.float(), beta=0, alpha=scale_factor)
+                attn_weights = attn_weights.reshape(bsz, num_heads, seq_len, seq_len)
```

Comment:
This can only be done if `autocast` was imported.

Reply:
What's the desired behaviour if torch<1.6? Should there be a check in the model `__init__` if `reorder_and_upcast_attn=True` that emits an error or warning?

Reply:
There can't be any autocast, so executing the code without the context manager will work fine.

Reply:
@lvwerra has added the version check!